### PR TITLE
[GEOT-5750] Add support for latest MySQL DB enhancements in jdbc-mysql module

### PR DESCRIPTION
**Pull request by nprigour**
Modifications of jdbc-mysql module to support latest MySQL DB  enhancements (version 5.6 and above)
This include mainly support for precise object shape spatial computations (use of ST_xxx() operations). 

In order to be backward compatible support is enabled by appropriate Map parameter which can be set during creation of JDBCDataStore or if not set it is computed based on MySQL DB version. 
 
Signed-off-by: Nikolaos Pringouris <nprigour@gmail.com>

### CODE REVIEW

**Comment by aaime** [comment]
Hi, thank you for reaching out, it's good to see there is still someone caring for mysql :-)
According to our procedure (GitHub linked to them when you opened the pull request), you should add:
- Send a contribution agreement to OSGeo (if you already did, my apologies, I don't remember everybody)
- Add tests verifying your changes work as expected against the desired version of mysql (and are not run otherwise, if the version is not recent enough).

Mind, the Travis build won't test databases, and main build server does test mysql but I don't know which version is used there. @jodygarnett or @tbarsballe do you know?

**Comment by tbarsballe** [comment]
Looks like the test MySQL version is 5.5.41. 
I'll see if we are able to update that version.

**Comment by nprigour** [comment]
Hi @aaime,

- I have already send a signed contribution agreement to info@osgeo.org since 3/3/2016 (after a discussion we had in the context of https://osgeo-org.atlassian.net/browse/GEOT-5350 asking me to become a maintainer of the mysql-jdbc plugin. By the way the issue  is addressed by the present pull request but also much more enhancements) but never received a reply on my email.

- I will have a look regarding the checks. Note however that the default behavior of the mysql datastore remains unaffected. The enhanced spatial support is enabled only if a proper parameter is set or a Mysql version > 5.6 is detected.

@tbarsballe 

- if you update the mysql version please go for the latest version that is 5.7.x.

**Comment by aaime** [comment]
Hi @nprigour there is nothing on file OSGeo wise. Unfortunately we do not get feedback from OSGeo, can I ask you to send the agreement again cc'ing me in the mail?

**Comment by nprigour** [comment]
Test cases added to verify proper setting of enhanced spatial support flag based on MySQL version.
The pull request also fixes [GEOT-5350] (https://osgeo-org.atlassian.net/browse/GEOT-5350)

**Comment by nprigour** [comment]
Hi @aaime, @tbarsballe
what is the status of this pull request and the pending issues to proceed with it (reception of agreement, installation of MySQL 5.7)?

**Comment by tbarsballe** [comment]
Our MySQL test database will be undergoing server migrations in the near future, and wont be upgraded until after that occurs. Unfortunately, I cannot give a timeframe for when that will happen.

**Comment by aaime** [comment]
I installed MySQL 5.7 locally for the sole purpose of reviewing this pull request (and the other one).
However, it seems I've hit a non starter, there are 28 tests failing on the current code base, probably due to changes in MySQL behavior.

I know pretty much nothing about MySQL and have not used it for ages now, but reviewing will be difficult if I don't start with a suite fully passing.

The first issue is that BBOX computation is not getting back a polygon anymore in some cases, but lines or points (if the source geometry is of that type). No big deal, I've changed the code to work with any geometry.

The second issue is that the setup for date based tests is not working anymore:

````
@Override
    protected void createDateTable() throws Exception {
        run( "CREATE TABLE dates (d DATE, dt DATETIME, t TIME)");
        
        run( "INSERT INTO dates VALUES (" +
                "STR_TO_DATE('2009-06-28', '%Y-%m-%d'), " +
                "STR_TO_DATE('2009-06-28 15:12:41', '%Y-%m-%d %H:%i:%s')," +
                "STR_TO_DATE('15:12:41', '%H:%i:%s')  )");
        
        run( "INSERT INTO dates VALUES (" +
                "STR_TO_DATE('2009-01-15', '%Y-%m-%d'), " +
                "STR_TO_DATE('2009-01-15 13:10:12', '%Y-%m-%d %H:%i:%s')," +
                "STR_TO_DATE('13:10:12', '%H:%i:%s')  )");
        
        run( "INSERT INTO dates VALUES (" +
                "STR_TO_DATE('2009-09-29', '%Y-%m-%d'), " +
                "STR_TO_DATE('2009-09-29 17:54:23', '%Y-%m-%d %H:%i:%s')," +
                "STR_TO_DATE('17:54:23', '%H:%i:%s')  )");

    }
````

Fails with:

````
java.sql.SQLException: Incorrect datetime value: '15:12:41' for function str_to_date
````
However according to the documentation it would seem this to be valid:
https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_str-to-date

but indeed it fails in a console:

````
select STR_TO_DATE('17:54:23', '%H:%i:%s') AS test;
+------+
| test |
+------+
| NULL |
+------+
1 row in set, 1 warning (0,00 sec)
````

Anyone has a clue as to how to handle this one?

**Comment by nprigour** [comment]
Hi @aaime,

What exact version of MySQL5.7 are you using. I have checked all the above statements in my local version of MySQL (v5.7.20-log win_64) and it seems that all the above statements you are mentioning in your comment are executed succesfully without any issues.

**Comment by aaime** [comment]
Here is what I have (out of a random internet search on getting the version, if you have a better/more detailed command let me know):

````
 SHOW VARIABLES LIKE "%version%";
+-------------------------+-------------------------+
| Variable_name           | Value                   |
+-------------------------+-------------------------+
| innodb_version          | 5.7.20                  |
| protocol_version        | 10                      |
| slave_type_conversions  |                         |
| tls_version             | TLSv1,TLSv1.1           |
| version                 | 5.7.20-0ubuntu0.16.04.1 |
| version_comment         | (Ubuntu)                |
| version_compile_machine | x86_64                  |
| version_compile_os      | Linux                   |
+-------------------------+-------------------------+
8 rows in set (0,00 sec)
````

**Comment by aaime** [comment]
Doh! It's in the docs, "If the NO_ZERO_DATE or NO_ZERO_IN_DATE SQL mode is enabled, zero dates or part of dates are disallowed. In that case, STR_TO_DATE() returns NULL and generates a warning:"
Check also here (both this variable and the next one):
https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_no_zero_date

**Comment by aaime** [comment]
@nprigour I've made this pull request to have all tests pass on my local 5.7, could you double check it results in a full pass on your end too?

https://github.com/geotools/geotools/pull/1731

If you have never run the tests, try running them once, a ``.geootols/mysql.properties.sample`` will be created in your home directory, remove the ``.sample`` from the extension, fix the contents according to your system, and then run the tests again. Locally for me it pumps for some 30 seconds and gives me 225 passed tests, 3 ignored (meaning the module is lacking around 100 tests that other databases have and that the MySQL module fails to extend and provide an implementation for... that's poor QA... )

**Comment by nprigour** [comment]
Hi @aaime,
The version of MySQL you are using is similar to mine but for the linux env. Concerning the checks:
I checkout your pull request (#1731), run MySQL checks (225 in total) and all were executed succesfully.
I also merge it with the current pull request and all checks (227 in total) were also executed succesfully.

**Comment by aaime** [comment]
@nprigour thanks for checking, I have merged that pull request.
I'm out of office for the week, in the meantime could you rebase this pull request and fix the conflicts?

File: build/maven/jar-collector/pom.xml [hunk 1]
```diff
@@ -16,7 +16,7 @@
   <parent>
     <groupId>org.geotools.maven</groupId>
     <artifactId>maven</artifactId>
-    <version>19-SNAPSHOT</version>
+    <version>18-SNAPSHOT</version>
```
Comment (bradh) [inline build/maven/jar-collector/pom.xml]: This doesn't look right for master. Why are you making this module depend on an earlier version of the parent?

File: build/release/build_release.sh [hunk 2]
```diff
@@ -127,9 +123,9 @@ ant -f rename.xml
 popd > /dev/null
 
 # build the release
-if [ "$SKIP_BUILD" != true ]; then
+if [ -z $SKIP_BUILD ]; then
   echo "building release"
-  export MAVEN_OPTS="-Xmx2048m"
+  export MAVEN_OPTS="-Xmx1024m"
```
Comment (bradh) [inline build/release/build_release.sh]: Why? This is a project wide setting.

File: build/release/properties [hunk 3]
```diff
@@ -21,7 +21,7 @@ if [ -z $DIST_PATH ]; then
 fi
 
 if [ -z $DIST_HOST ]; then
-  DIST_HOST=build.geoserver.org
+  DIST_HOST=ares.boundlessgeo.com
```
Comment (bradh) [inline build/release/properties]: This doesn't look right.

**Comment by nprigour** [comment]
Hi @bradh,
I tried to rebase and push to this pull request branch after synching my forked  geotools master but probably I messed everything up. So I created a revert commit and forced pushed it to this branch. 
It should be OK now but I would be glad if somebody can explain how to rebase and update a pull request that is several commits behind the master

**Comment by aaime** [comment]
@nprigour indeed while I could merge this on master, I would not know where to start to do a backport due to its current weird structure:

image

If you are interested in a backport please create a new pull request with a linear history (no merge commits), otherwise if you're happy with just having it on master, I'll be equally happy to hit the merge button :-p

**Comment by nprigour** [comment]
Hi @aaime,
when you refer to creation of a new pull request in order to be able to backport, what do you expect to be the starting point? The current master or the initial commit around 5 months ago. Would it suffice to just create a new pull request after updating my fork to be in sync with the geotools master?

**Comment by aaime** [comment]
@nprigour backport is easier if the pull request has a single commit based on the current master (so that I can simply cherry-pick it back)

**Comment by nprigour** [comment]
see #1739

**Pull request closed without merging**